### PR TITLE
Propagate ClusterID to Cloud Map & Endpoints

### PR DESCRIPTION
**Pull request by matthewgoodman13**
*Affiliated with [Issue 47](https://github.com/aws/aws-cloud-map-mcs-controller-for-k8s/issues/47)*

*TLDR:*
ClusterID getting exported to Cloud Map & getting propagated to endpoints as an `CLUSTER_ID` attribute

*Description of changes:*
- Added the ClusterProperty CRD (from new group `about.k8s.io`)
- In `serviceexport_controller.go`, we now watch for changes in the ClusterProperty attribute, and when there is a change, we reconcile all endpoints. This works for adding and removing of a ClusterProperty object. 
    - This can be tested by `kubectl apply -f config/samples/about_v1alpha1_clusterproperty.yaml`
- [Update 1] If no ClusterId is defined - an error will now be thrown. Service Export will not be created.
- [Update 2 - b35bc75] ClustersetId added. This id is sent in DiscoverInstances as a QueryParameter to only return endpoints within the clusterset. More unit tests added.
- [Update 3] ClusterUtils added. **Integration tests will fail until next PR.**

*Notes:*
- This is a minimal PR. Main goal was to add the new about api. ~~No error handling introduced yet and no errors will be thrown yet for no ClusterID being defined other than when the service is being exported. No unit tests in this PR either.~~
- Adding / removing the ClusterProperty (id.k8s.io) will add/remove the `CLUSTER_ID` attribute in the endpoint
- In `suite_test.go`, changes needed to be made to not include the clusterproperty CRD for testing. Special tests will need to be constructed to manually add the annotation to be allowed to use '.k8s.io' suffix.

*Testing:*
- manual testing of the controller
- `make test` runs without error
- [Update 1] tests updated to include ClusterID. Unit test added for no cluster id - expect
- [Update 2] more unit tests added for Clusterid and ClustersetId
- [Update 3] Integration tests will fail until next PR

### CODE REVIEW

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 1]
```diff
@@ -241,6 +258,12 @@ func (r *ServiceExportReconciler) extractEndpoints(ctx context.Context, svc *v1.
 					if version.GetVersion() != "" {
 						attributes[K8sVersionAttr] = version.PackageName + " " + version.GetVersion()
 					}
+
+					// Apply clusterID as endpoint attribute if it exists
+					if clusterId.Spec.Value != "" {
+						attributes[ClusterIdAttr] = clusterId.Spec.Value
```
Comment (astaticvoid) [inline pkg/controllers/multicluster/serviceexport_controller.go:264]: Have we considered adding this to our endpoint model as a field?

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 2]
```diff
@@ -225,6 +236,12 @@ func (r *ServiceExportReconciler) extractEndpoints(ctx context.Context, svc *v1.
 		return nil, err
 	}
 
+	clusterId := &aboutv1alpha1.ClusterProperty{}
```
Comment (astaticvoid) [inline pkg/controllers/multicluster/serviceexport_controller.go:239]: Since the parent Reconcile function already retrieved this, should we pass the value into the extractEndpoints function and save some unit test coverage percentage?

File: config/samples/about_v1alpha1_clusterproperty.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# An example object of `id.k8s.io ClusterProperty` 
+# using a kube-system ns uuid as the id value:
```
[the remark this answers is not in the document]
Comment (matthewgoodman13) [reply config/samples/about_v1alpha1_clusterproperty.yaml:2]: updated comment.

File: pkg/controllers/multicluster/serviceexport_controller.go (hunk 2 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:239]: sure!

File: pkg/controllers/multicluster/serviceexport_controller.go (hunk 1 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:264]: In my opinion, I like it as an attribute. The other existing fields (Id, IP, Ports) seems like more intrinsic properties that would belong to an Endpoint. ClusterID seems more like an attribute or additional useful information to have when getting Endpoint information. Im happy with either. What do we think?

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 4]
```diff
@@ -286,6 +310,26 @@ func (r *ServiceExportReconciler) endpointSliceEventHandler() handler.MapFunc {
 	}
 }
 
+func (r *ServiceExportReconciler) clusterPropertyEventHandler() handler.MapFunc {
+	// Return reconcile requests for all services
+	return func(object client.Object) []reconcile.Request {
+		services := &v1.ServiceList{}
```
[the remark this answers is not in the document]
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:316]: done.

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 5]
```diff
@@ -69,6 +72,15 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	// Mark ServiceExport to be deleted, which is indicated by the deletion timestamp being set.
 	isServiceExportMarkedForDelete := serviceExport.GetDeletionTimestamp() != nil
 
+	clusterId := &aboutv1alpha1.ClusterProperty{}
+	err := r.Client.Get(ctx, client.ObjectKey{Name: ClusterIdName}, clusterId)
+	if err != nil {
+		r.Log.Error(err, "error fetching ClusterId")
+		return ctrl.Result{}, err
+	} else {
+		r.Log.Debug("ClusterID found", "ClusterID", clusterId)
```
[the remark this answers is not in the document]
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:81]: Refactored code block to `clusterProperty=...` to make it clearer. 
Then outputting: `r.Log.Debug("ClusterProperty with ClusterID found", "ClusterID", clusterProperty.Spec.Value)` which outputs `DEBUG   controllers.ServiceExport       ClusterProperty with ClusterID found    {"ClusterID": "sample-mcs-clusterid"}` in console.

File: pkg/controllers/multicluster/serviceexport_controller.go (hunk 1 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:259]: Changed to field.

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 6]
```diff
@@ -286,6 +310,26 @@ func (r *ServiceExportReconciler) endpointSliceEventHandler() handler.MapFunc {
 	}
 }
 
+func (r *ServiceExportReconciler) clusterPropertyEventHandler() handler.MapFunc {
+	// Return reconcile requests for all services
+	return func(object client.Object) []reconcile.Request {
+		services := &v1.ServiceList{}
+		if err := r.Client.List(context.TODO(), services); err != nil {
+			r.Log.Error(err, "error listing services")
+			return nil
```
[the remark this answers is not in the document]
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:319]: If for some reason `r.Client.List(serviceExports)` fails, then the r.Log.Error will display and the return nil will simply just quit the event handler method and not force a reconcile. No other unexpected behavior happens when nil is returned (even to its parent methods that called it) - that is, no other errors are returned. As we have it right now, there is no retry.

File: pkg/controllers/multicluster/serviceexport_controller.go (hunk 1 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:259]: As per @fredjywang 's design review meeting - confirmed to go with field. @fredjywang will also add new field for `ServiceType`

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 7]
```diff
@@ -28,6 +29,8 @@ import (
 )
 
 const (
+	ClusterIdName             = "id.k8s.io"
```
[the remark this answers is not in the document]
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:32]: Added clustersetid in the same manner as clusterid. Both clustersetid and clusterid must be defined to export a service from a cluster. When calling DiscoverInstances, we now use the clustersetid as a queryparameter to only return endpoints that are in the same clusterset

File: integration/janitor/runner/main.go [hunk 8]
```diff
@@ -9,12 +9,13 @@ import (
...
+	if len(os.Args) != 3 {
+		fmt.Println("Expected single namespace name and clustersetId arguments")
 		os.Exit(1)
 	}
 
 	j := janitor.NewDefaultJanitor()
 	nsName := os.Args[1]
-	j.Cleanup(context.TODO(), nsName)
+	clustersetId := os.Args[2]
+	j.Cleanup(context.TODO(), nsName, clustersetId)
```
Comment (matthewgoodman13) [inline integration/janitor/runner/main.go:20]: Since the discover instances that is used in the integration requires the new `clustersetId` param, I had to add this to main.go for the integration tests. @fredjywang

File: integration/shared/scenarios/export_service.go [hunk 9]
```diff
@@ -82,7 +82,7 @@ func (e *exportServiceScenario) Run() error {
 
 	return wait.Poll(defaultScenarioPollInterval, defaultScenarioPollTimeout, func() (done bool, err error) {
 		fmt.Println("Polling service...")
-		cmSvc, err := e.sdClient.GetService(context.TODO(), e.expectedSvc.Namespace, e.expectedSvc.Name)
+		cmSvc, err := e.sdClient.GetService(context.TODO(), e.expectedSvc.Namespace, e.expectedSvc.Name, "")
```
Comment (matthewgoodman13) [inline integration/shared/scenarios/export_service.go:85]: Unsure what to do here.

Comment (astaticvoid) [reply integration/shared/scenarios/export_service.go:85]: Tests will currently succeed since clusterset id is presumably being exported as blank during test. Once cluster id and cluster set id is required kind/eks test setups will need to set this in the test clusters. The `exportServiceScenario` object should get a clusterset id field and pass that value to `GetService`.

File: pkg/cloudmap/api.go [hunk 10]
```diff
@@ -113,12 +113,15 @@ func (sdApi *serviceDiscoveryApi) GetServiceIdMap(ctx context.Context, nsId stri
...
 
-func (sdApi *serviceDiscoveryApi) DiscoverInstances(ctx context.Context, nsName string, svcName string) (insts []types.HttpInstanceSummary, err error) {
+func (sdApi *serviceDiscoveryApi) DiscoverInstances(ctx context.Context, nsName string, svcName string, clustersetId string) (insts []types.HttpInstanceSummary, err error) {
 	out, err := sdApi.awsFacade.DiscoverInstances(ctx, &sd.DiscoverInstancesInput{
 		NamespaceName: aws.String(nsName),
 		ServiceName:   aws.String(svcName),
 		HealthStatus:  types.HealthStatusFilterAll,
 		MaxResults:    aws.Int32(1000),
+		QueryParameters: map[string]string{
+			model.ClustersetIdAttr: clustersetId,
```
Comment (runakash) [inline pkg/cloudmap/api.go:123]: needs to be `aws.String(clustersetId)`

File: pkg/controllers/multicluster/cloudmap_controller.go [hunk 11]
```diff
@@ -53,25 +54,33 @@ func (r *CloudMapReconciler) Start(ctx context.Context) error {
 
 // Reconcile triggers a single reconciliation round
 func (r *CloudMapReconciler) Reconcile(ctx context.Context) error {
+	clusterPropertyForClustersetId := &aboutv1alpha1.ClusterProperty{}
+	err := r.Client.Get(ctx, client.ObjectKey{Name: ClustersetIdName}, clusterPropertyForClustersetId)
```
Comment (runakash) [inline pkg/controllers/multicluster/cloudmap_controller.go:58]: Shouldn't we also check for existence of clusterId?

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 12]
```diff
@@ -69,8 +72,32 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		err := r.Client.Get(ctx, client.ObjectKey{Name: ClusterIdName}, clusterPropertyForClusterId)
+		if err != nil {
+			r.Log.Error(err, "error fetching ClusterProperty with name "+ClusterIdName)
+			return ctrl.Result{}, err
+		} else {
+			r.Log.Debug("ClusterProperty with ClusterId found", "ClusterId", clusterPropertyForClusterId.Spec.Value)
+		}
+
+		// Fetch ClusterProperty with name "clusterset.k8s.io"
+		err = r.Client.Get(ctx, client.ObjectKey{Name: ClustersetIdName}, clusterPropertyForClustersetId)
```
Comment (runakash) [inline pkg/controllers/multicluster/serviceexport_controller.go:88]: Instead of making two API calls you can fetch ClusterProperty list, and iterate over them to verify both clustersetId and clusterId objects. Also, consider moving this duplicated logic to utils as this is used in both the controllers.

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 13]
```diff
@@ -164,8 +191,8 @@ func (r *ServiceExportReconciler) handleUpdate(ctx context.Context, serviceExpor
 	return ctrl.Result{}, nil
 }
 
-func (r *ServiceExportReconciler) createOrGetCloudMapService(ctx context.Context, service *v1.Service) (*model.Service, error) {
-	cmService, err := r.CloudMap.GetService(ctx, service.Namespace, service.Name)
+func (r *ServiceExportReconciler) createOrGetCloudMapService(ctx context.Context, service *v1.Service, clustersetId string) (*model.Service, error) {
+	cmService, err := r.CloudMap.GetService(ctx, service.Namespace, service.Name, clustersetId)
```
Comment (runakash) [inline pkg/controllers/multicluster/serviceexport_controller.go:195]: As `clustersetId` is global unique variable, it makes sense to set this globally somewhere instead of passing it around.

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 14]
```diff
@@ -286,6 +322,26 @@ func (r *ServiceExportReconciler) endpointSliceEventHandler() handler.MapFunc {
 	}
 }
 
+func (r *ServiceExportReconciler) clusterPropertyEventHandler() handler.MapFunc {
+	// Return reconcile requests for all services
```
Comment (runakash) [inline pkg/controllers/multicluster/serviceexport_controller.go:326]: Please update the comment

File: pkg/controllers/multicluster/suite_test.go [hunk 15]
```diff
@@ -37,8 +37,14 @@ var _ = BeforeSuite(func() {
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
 	By("bootstrapping test environment")
+
+	crds := []string{
+		filepath.Join("..", "..", "..", "config", "crd", "bases", "multicluster.x-k8s.io_serviceexports.yaml"),
+		filepath.Join("..", "..", "..", "config", "crd", "bases", "multicluster.x-k8s.io_serviceimports.yaml"),
+	}
+
```
Comment (runakash) [inline pkg/controllers/multicluster/suite_test.go:45]: should we have suite_test.go to verify clusterproperty CRDs?

**Review by runakash — CHANGES_REQUESTED** [review]

File: pkg/cloudmap/api.go (hunk 10 above)
Comment (matthewgoodman13) [reply pkg/cloudmap/api.go:123]: The function returns the wrong type expected by QueryParameters and causes type errors. The aws functions return pointer values. Also tried with StringMap(queryParams). Also tried with both combined.

Looking into the types file for DiscoverInstancesInput, the expected type is `map[string]string
` and no pointers for the map's strings either. [See here.](https://github.com/aws/aws-sdk-go-v2/blob/98519e67e1d575b00ed51b831a877a4b9bbf6a43/service/servicediscovery/api_op_DiscoverInstances.go#L70)

Errors:
`cannot use aws.String(clustersetId) (value of type *string) as string value in map literal`
`cannot use aws.StringMap((map[string]string literal)) (value of type map[string]*string) as map[string]string value in struct literal`

File: pkg/controllers/multicluster/serviceexport_controller.go (hunk 13 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:195]: Added as global variables with additional `refresh` variable. Could of used caching but seems excessive for the use case.

File: pkg/controllers/multicluster/suite_test.go (hunk 15 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/suite_test.go:45]: As per the original comment up top: 
_In suite_test.go, changes needed to be made to not include the clusterproperty CRD for testing. Special tests will need to be constructed to manually add the annotation to be allowed to use '.k8s.io' suffix._

Comment (runakash) [reply pkg/controllers/multicluster/suite_test.go:45]: Can you elaborate further? What changes are needed.

File: pkg/cloudmap/api.go (hunk 10 above)
Comment (astaticvoid) [reply pkg/cloudmap/api.go:123]: FYI this was a change between sdk v1 and v2. APIGW uses v1 to call DiscoverInstances which can result in null query parameter values.

File: pkg/model/types.go [hunk 16]
```diff
@@ -123,6 +135,16 @@ func servicePortFromAttr(attributes map[string]string) (port Port, err error) {
 	return port, err
 }
 
+func clusterIdFromAttr(attributes map[string]string) (clusterId string, err error) {
```
Comment (astaticvoid) [inline pkg/model/types.go:138]: What is the benefit of creating wrapper functions for these parse actions instead of direct calls to `removeStringAttr()` like getting `endpoint.IP` does?

File: pkg/controllers/multicluster/suite_test.go (hunk 15 above)
Comment (matthewgoodman13) [reply pkg/controllers/multicluster/suite_test.go:47]: [See issue. ](https://github.com/aws/aws-cloud-map-mcs-controller-for-k8s/issues/172)

File: pkg/controllers/multicluster/serviceexport_controller.go [hunk 17]
```diff
@@ -33,12 +34,18 @@ const (
 	EndpointSliceServiceLabel = "kubernetes.io/service-name"
 )
 
+var (
+	clusterId    string
+	clusterSetId string
```
Comment (runakash) [inline pkg/controllers/multicluster/serviceexport_controller.go:39]: why is this required?

Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:39]: Not required - but was an attempt to make it cleaner to access the ids throughout the controller without having to pass around the clusterId, clusterSetId I retrieve on lines 65 & 70. Another option is to call the getClusterId function again in extractEndpoints.

Comment (runakash) [reply pkg/controllers/multicluster/serviceexport_controller.go:39]: > Another option is to call the getClusterId function again in extractEndpoints.

Let's do that. We do have context in the controller.

Comment (matthewgoodman13) [reply pkg/controllers/multicluster/serviceexport_controller.go:39]: done.

**Review by runakash — APPROVED** [review]
